Stop AIs from shooting at dead players + more

toberge opened the pull request:
Also closes #533

toberge committed:
Stop AIs from shooting at dead players

toberge committed:
Make AIs face the direction they are shooting

toberge committed:
Make AIs target the player that shoots at them

toberge committed:
Avoid exception from AI targeting w/o other players left

toberge committed:
Make seeker body usable by AIs

toberge committed:
Make tether body usable by AIs

toberge committed:
Make AIs shoot explosive barrels close to target

Fueredoriku commented on Assets/Scripts/Gamestate/AIManager.cs:
```diff
@@ -195,32 +210,20 @@ private IEnumerator LookForTargets()
+        // TODO the ammo box stuff!
+        var hasAmmoBoxBody = ammoBoxCollector.CanReload;
+        var isOutOfAmmo = gunController && gunController.stats.Ammo < 1;
+        if (hasAmmoBoxBody && isOutOfAmmo)
```
This should also be called when the AI is in close proximity mode! If the AI depletes its ammo and is in close range it'll soft lock itself not being able to shoot and unwilling to find new ammo

toberge replied:
Reworked so it happens when they are in close proximity, _without_ messing up their trajectory if they are in the air

Fueredoriku replied:
Nice!

Fueredoriku commented on Assets/Scripts/Gamestate/AIManager.cs:
```diff
@@ -237,10 +240,23 @@ private void FindPlayers()
+                && Vector3.Distance(closestExplodingBarrel.transform.position, transform.position) > closestExplodingBarrel.Radius + 1;
+            if (isBarrelSafeToShoot)
+            {
+                Debug.Log($"Found exploding barrel at {closestExplodingBarrel.transform.position} >:)", closestExplodingBarrel);
```
remove debug log?  🙏

toberge replied:
Removed 😔

Fueredoriku reviewed (changes requested):
Great work, the AIs become suprisingly more fun to play against when they can use seeker and tether too!
 
I have identified three issues that need fixing:
- Ammoboxes ignored if AI has entered a close range state
- Tether may never be replugged
(I suggest we fix the tether problem by forcing the AIs to always aim straight down when planting the plug)
![tethersoftlock](https://github.com/user-attachments/assets/2d4e6fee-3e87-4eb3-baca-fbe6ab61047b)



- Also it seems that if the player is gone, the AIs may sometimes choose to stand still at some place and shoot the remaining targets from that single spot only. To replicate this problem, make one or more of the AIs target you, die, then spectate and there is a good chance that they may get stuck standing in a specific place. See one of the comments I made for an example.

The rest looks good here, so if we fix these issues this should be good to go!

Fueredoriku commented on Assets/Scripts/Gamestate/AIManager.cs:
```diff
@@ -260,6 +276,55 @@ private void FindPlayers()
+                closestPlayer = player.AiTarget.transform;
+                ShootingTarget = player.AiAimSpot;
+                continue;
+            }
```
There seems to be a problem with something inside the areOnlyAiPlayersLeft, making it so the AIs can choose to stand still firing at eachother from places they can't hit eachother, thus softlocking the game

Here is an example spot from The Pit
![image](https://github.com/user-attachments/assets/929b9aa3-f4f5-4f3d-9436-320bf594c88d)

toberge replied:
Fixed. Removed the ShootingTarget and continue; lines, instead setting DestinationTarget.

toberge committed:
Fix targeting when only AIs are left

toberge committed:
Prevent dead player transform from lingering in AI

toberge commented:
> Great work, the AIs become suprisingly more fun to play against when they can use seeker and tether too!
> 
> I have identified three issues that need fixing:
> 
> * Ammoboxes ignored if AI has entered a close range state

Fixed, see other comment

>  * Tether may never be replugged
>     (I suggest we fix the tether problem by forcing the AIs to always aim straight down when planting the plug)

Fixed as suggested


> * Also it seems that if the player is gone, the AIs may sometimes choose to stand still at some place and shoot the remaining targets from that single spot only. To replicate this problem, make one or more of the AIs target you, die, then spectate and there is a good chance that they may get stuck standing in a specific place. See one of the comments I made for an example.

Fixed. They don't instantly set ShootingTarget here anymore, instead DestinationTarget so that they walk to their enemy first.

toberge committed:
Fix ai movement not being enabled on knockback

Fueredoriku commented on Assets/Scripts/Gamestate/AIManager.cs:
```diff
@@ -186,61 +227,78 @@ private IEnumerator LookForTargets()
+                    // TODO the shooting target going missing seems to be the cause for 
+                    ShootingTarget = closestExplodingBarrel.transform;
+                    // TODO unsub?
+                    closestExplodingBarrel.GetComponent<HealthController>().onDeath += HandleExplodedBarrel;
```
Maybe unsubscribe at the HandleExplodedBarrel function itself? Might be nice in order to minimize memory leakage

toberge replied:
Thought I did that already but must have forgor 💀

toberge replied:
Nope, I already did every single thing that the TODO comments around here were talking about. Removed the superfluous comments now 🙃

Fueredoriku reviewed (approved):
Great work!! 👍

The AIs seem a lot more... sane now 😅

I left a small comment but other than that I have no more feedback, so I'll leave an approve for you to use at your discretion
